Vault Agent Cache

vishalnayak opened the pull request.

briankassouf commented on command/agent/cache/lease_cache.go:
```diff
@@ -0,0 +1,743 @@
+
+	for {
+		select {
+		case <-ctx.Done():
```
Is one of these cases handling full shutdown of Agent?

vishalnayak replied:
Added comments for all the select cases. Yes, this first case of `ctx.Done()` should cover shutdown as well.

briankassouf commented on command/agent/cache/lease_cache.go:
```diff
@@ -0,0 +1,743 @@
+	})
+}
+
+func (c *LeaseCache) handleCacheClear(ctx context.Context, clearType string, clearValues ...interface{}) error {
```
The `...interface{}` feels a little too flexible and easy to mess up, for instance if you are planning on clearing all you'd need to provide dummy values otherwise you will get caught in the below check. I wonder if it makes sense to provide a struct here instead. It would contain the Type, namespace, and Value? or instead of type it could be IndexType?

jefferai replied:
Yeah I agree. I think this would be better as something like:
```go
type ClearInfo struct {
    RequestPaths []string
    Tokens []string
    TokenAccessors []string
    LeasePrefixes []string
    Everything bool
}
```

calvn replied:
I implemented the change using this struct:
```
type clearOptions struct {
	ClearType     string
	RequestPath   string
	Namespace     string
	Token         string
	TokenAccessor string
	Lease         string
}
```
I can probably get rid of `ClearType` since the types are technically present as fields, but how can I ensure that only one type is provided (with the exception of `RequestPath`, which can optionally include a non-empty `Namespace` value) without having a particular preference if more than one is provided. The ideal thing would be to not accept more than one. If `Token` and `Lease` where both provided, for instance, which one should handleCacheClear accept (again, the ideal thing would be to error out and say that only one field can be provided, but not sure how an "ensure only one field is set" can be coded without preference bias)?

jefferai replied:
Why is it ideal to have only one? Why can't multiple be provided, and why can't multiple values of several of those fields be provided at once?

briankassouf commented on command/agent/cache/lease_cache.go:
```diff
@@ -0,0 +1,743 @@
+		token, ok := jsonBody["token"]
+		if !ok {
+			return false, fmt.Errorf("failed to get token from request body")
+		}
```
For safety can we verify token is a string here with an `, ok`?

vishalnayak replied:
Fixed.

briankassouf commented on command/agent/cache/lease_cache.go:
```diff
@@ -0,0 +1,743 @@
+		accessor, ok := jsonBody["accessor"]
+		if !ok {
+			return false, fmt.Errorf("failed to get accessor from request body")
+		}
```
Same, let's verify string

vishalnayak replied:
Fixed.

briankassouf commented on command/agent/cache/lease_cache.go:
```diff
@@ -0,0 +1,743 @@
+			if index.Lease != "" {
+				index.RenewCtxInfo.CancelFunc()
+			} else {
+				close(index.RenewCtxInfo.DoneCh)
```
Why are we closing this? Wouldn't this evict all child tokens from cache?

vishalnayak replied:
This is not cancelling the context and hence the child tokens will not be evicted. It is closing the channel that returns the renewer goroutine. So, it only removes a single item.

briankassouf replied:
But which item are we removing? we don't want to stop renewing child tokens, right?

vishalnayak replied:
If an intermediate token is revoke, its parent and parents lease should be intact. The child tokens and their respective leases should be intact. Only the revoked token and the leases belonging to the revoked token should be gone. I'll definitely need a test for this case which I am working on, but the code seems to be doing the right thing. The child tokens of the tokens being revoked are being manually walked over and cancelled individually, and then the renewer of the revoked token is inform to return without cancelling its context. That way the children contexts are unaffected.

vishalnayak replied:
Added a test to cover this case. It's working as expected.

briankassouf commented on command/agent/cache/lease_cache.go:
```diff
@@ -0,0 +1,743 @@
+		prefix := strings.TrimPrefix(path, vaultPathLeaseRevokeForce)
+		// Get all the cache indexes that use the request path containing the
+		// prefix and cancel the renewer context of each.
+		indexes, err := c.db.GetByPrefix("request_path", namespace, "/v1"+prefix)
```
I think we'd want to lookup by lease prefix, not request_path?

vishalnayak replied:
@calvn can answer better, but I _think_ this was intended. I think the reason for this being like the way it is, is that only the `request_path` is indexed along with namespace and `lease` is not. The fact that request path is always embedded in the lease id is utilized in conjunction with namespace to get the desired effect.

calvn replied:
Now that the namespace is returned along with the lease ID in https://github.com/hashicorp/vault/pull/6212, we might be able to switch the index to a compound index, similar to `request_path` indexer.

briankassouf commented on command/agent/cache/lease_cache.go:
```diff
@@ -0,0 +1,743 @@
+		prefix := strings.TrimPrefix(path, vaultPathLeaseRevokePrefix)
+		// Get all the cache indexes that use the request path containing the
+		// prefix and cancel the renewer context of each.
+		indexes, err := c.db.GetByPrefix("request_path", namespace, "/v1"+prefix)
```
Should this be lease ID prefix, not request path?

vishalnayak replied:
Same as above.

vishalnayak commented on command/agent.go:
```diff
@@ -332,10 +335,40 @@ func (c *AgentCommand) Run(args []string) int {
 
+	// Parse agent listener configurations
+	var listeners []net.Listener
+	if len(config.Cache.Listeners) != 0 {
```
We should not do anything if `cache` block is not configured and if there no listeners listed in the `cache` block. Fixed.

vishalnayak commented on command/agent.go:
```diff
@@ -332,10 +335,40 @@ func (c *AgentCommand) Run(args []string) int {
+
+	// Start listening to requests
+	err = cache.Run(ctx, &cache.Config{
+		Token:            c.client.Token(),
```
You are right. Updated it to dynamically fetch it from the agent command's client.

vishalnayak commented on command/agent/cache/api_proxy.go:
```diff
@@ -0,0 +1,58 @@
+		return nil, err
+	}
+	client.SetToken(req.Token)
+	client.SetHeaders(req.Request.Header)
```
At this point, yes. The main reason for not adding a configuration option is that the `vault agent` is already using the env vars to talk to Vault. Introducing Vault addresses with `cache` block felt weird. Nevertheless, we can think a bit more on that decide the best path forward. Shouldn't be a hard change.

vishalnayak commented on command/agent/cache/api_proxy.go:
```diff
@@ -0,0 +1,58 @@
+	if err != nil {
+		ap.logger.Error("failed to read request body", "error", err)
+		return nil, err
+	}
```
We shouldn't be closing the response body here as it will be read out up in the http layer. I saw that it wasn't getting closed there as well. Added that as a `defer` there. Kindly take a look at the recent commit that does that.

briankassouf replied:
But we are creating new read closers, so we are closing the new ones but not the existing one?

vishalnayak replied:
Possibly. I've fixed this in c4c930ad52dc6e238d6dea3fa8bbce3b6dd1833f. Kindly take a look.

michelvocks commented on command/agent/cache/handler.go:
```diff
@@ -0,0 +1,127 @@
+	Logger           hclog.Logger
+}
+
+func Run(ctx context.Context, config *Config) error {
```
Can `config` be `nil`? Maybe add a check like you did at https://github.com/hashicorp/vault/pull/6220/files#diff-703e1234dc09e66de48735d23f23732eR90

vishalnayak replied:
Fixed.

ncabatoff commented on command/agent/cache/lease_cache.go:
```diff
@@ -0,0 +1,746 @@
+	proxier       Proxier
+	logger        hclog.Logger
+	db            *cachememdb.CacheMemDB
+	rand          *rand.Rand
```
I was referring to the field 'rand'.  Just realized that wasn't clear from context.

vishalnayak replied:
Yes. Fixed.

ncabatoff commented on command/agent/cache/lease_cache.go:
```diff
@@ -0,0 +1,746 @@
+	logger        hclog.Logger
+	db            *cachememdb.CacheMemDB
+	rand          *rand.Rand
+	tokenContexts map[string]*ContextInfo
```
For that matter, why do we need this map?  Why not just always get the ContextInfo from Index.RenewCtxInfo and save some memory?

vishalnayak replied:
Probably we need one for the current state of things. For the question you posed as to whether this is needed at all, the idea for having a separate map was to ensure that there were placeholders for tokens that are not managed by the agent. For example, for tokens that were not created through the agent, there won't be entries in memdb and hence there won't by any ContextInfo. *But*, at some point we decided that we'll only manage leases that are resulted from the tokens that were created through cache. That decision probably obviates the need for this map. I'll think on this for a while, and if it makes sense to remove it, I will. Thanks for pointing this out!

vishalnayak replied:
Removed the map.

ncabatoff commented on command/agent/cache/lease_cache.go:
```diff
@@ -0,0 +1,762 @@
+		}
+
+		// Find out all the indexes that are directly tied to the revoked token
+		indexes, err := c.db.GetByPrefix(cachememdb.IndexNameToken.String(), token)
```
Why is this (and the call below) GetByPrefix instead of simply Get?

vishalnayak replied:
The goal of this call is to get all the entries in the cache that are attached to the token. While `GetByPrefix` returns all the attached entries, `Get` returns the first match.

ncabatoff replied:
Got it, thanks.

vishalnayak commented on command/agent/config/config_test.go:
```diff
@@ -10,6 +10,83 @@ import (
+func TestLoadConfigFile_AgentCache(t *testing.T) {
+	logger := logging.NewVaultLogger(log.Debug)
+
+	os.Setenv("TEST_AAD_ENV", "aad")
```
This is not required for this test. It is an environment variable expected by the `sink` of the `auto-auth` block. This environment variable is used to supply `additional authenticated data`. This test doesn't need this var. Thanks for catching. Removed.

vishalnayak commented on command/agent/cache/proxy.go:
```diff
@@ -0,0 +1,28 @@
+// SendResponse is the output from Proxier.Send.
+type SendResponse struct {
+	Response     *api.Response
+	ResponseBody []byte
```
Response and request bodies were read out multiple times by different layers, which was essentially "read the response body, do whatever with it and reset the body in the request/response in a readable state". Piggy backing the bodies alongside avoid some instances of that.

ncabatoff replied:
Makes sense, thanks for clarifying.

vishalnayak commented on command/agent/config/config_test.go:
```diff
@@ -10,6 +10,83 @@ import (
+				&Listener{
+					Type: "unix",
+					Config: map[string]interface{}{
+						"address":     "/Users/vishal/go/src/github.com/hashicorp/vault/socket",
```
Resultant of copy-pasta from my test config file. Fixed.

vishalnayak commented on command/agent/cache/cache_test.go:
```diff
@@ -0,0 +1,321 @@
+}
+`
+
+// testSetupClusterAndAgent is a helper func used to set up a test cluster and
```
Fixed.

vishalnayak commented on command/agent/cache/cachememdb/cache_memdb.go:
```diff
@@ -0,0 +1,239 @@
+							},
+						},
+					},
+					IndexNameToken.String(): &memdb.IndexSchema{
```
Done.

calvn commented on command/agent/cache/cachememdb/cache_memdb_test.go:
```diff
@@ -0,0 +1,286 @@
+	}
+}
+
+func TestCacheMemDB_Evict(t *testing.T) {
```
Added verification after evicition

calvn commented on command/agent/cache/cache_test.go:
```diff
@@ -0,0 +1,321 @@
+	token := resp.Auth.ClientToken
+	testClient.SetToken(token)
+
+	// Test on auth response by creating a child token
```
Done.

vishalnayak commented on command/agent/cache/lease_cache.go:
```diff
@@ -0,0 +1,743 @@
+	if !secret.Renewable && !tokenRenewable {
+		c.logger.Debug("secret not renewable, skipping addtion to the renewer")
+		return resp, nil
+	}
```
Done!

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/vault?pullRequest=6220)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/vault?pullRequest=6220)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)


---
**9** out of **10** committers have signed the CLA.

- [x] jefferai
- [x] vishalnayak
- [x] briankassouf
- [x] brianshumate
- [x] kalafut
- [x] tyrannosaurus-becks
- [x] roooms
- [x] meirish
- [x] varshavaradarajan
- [ ] dominic


---
**dominic** seems not to be a GitHub user.
You need a GitHub account to be able to sign the CLA. If you already have a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).

<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/vault?pullRequest=6220).</sub>

vishalnayak commented on command/agent/cache/handler.go:
```diff
@@ -0,0 +1,123 @@
+		if err != nil {
+			config.Logger.Error("failed to read request body")
+			respondError(w, http.StatusInternalServerError, errors.New("failed to read request body"))
+		}
```
All the cases should be covered by c4c930ad52dc6e238d6dea3fa8bbce3b6dd1833f.

jefferai commented on api/client.go:
```diff
@@ -366,6 +371,21 @@ func NewClient(c *Config) (*Client, error) {
+	// the HttpClient's transport to the corresponding socket dialer.
+	if strings.HasPrefix(c.Address, "unix://") {
+		socketFilePath := strings.TrimPrefix(c.Address, "unix://")
+		c.HttpClient = &http.Client{
```
By doing this you're overriding the values set in cleanhttp, plus any values set in DefaultClient, all of which are likely mostly still relevant here. Rather than set an entirely new client and transport, why not simply set a new DialContext on the existing ones?

vishalnayak replied:
Agreed! Maybe this was the problem I was running into earlier when HTTPS over sockets were throwing weird errors, which had lead me to force TLS disabling on unix sockets for the time being. I've re-enabled that.

vishalnayak replied:
Addressed in https://github.com/hashicorp/vault/pull/6237/files

jefferai commented on api/request.go:
```diff
@@ -85,7 +85,7 @@ func (r *Request) ToHTTP() (*http.Request, error) {
 }
 
-func (r *Request) toRetryableHTTP() (*retryablehttp.Request, error) {
+func (r *Request) ToRetryableHTTP() (*retryablehttp.Request, error) {
```
Please do not make this public.

jefferai replied:
(Use ToHTTP instead)

vishalnayak replied:
Fixed.

jefferai commented on command/agent.go:
```diff
@@ -332,10 +324,92 @@ func (c *AgentCommand) Run(args []string) int {
+
+		// Create a muxer and add paths relevant for the lease cache layer
+		mux := http.NewServeMux()
+		mux.Handle("/v1/agent/cache-clear", leaseCache.HandleCacheClear(ctx))
```
I thought the plan was to have different endpoints for clearing different parts of the cache (eventually). Given that, I'd put this behind a reasonable collection, e.g. /v1/agent/cache/clear-all.

I also wonder whether we need a /v1/ at all. If we don't use that for cache paths, we can automatically know that any path coming in starting with /v1/ is destined for Vault.

vishalnayak replied:
I removed the `/v1` prefix for the API as it makes total sense and the change was easy as well. As to the other point on breaking the API into parts, I'll wait for other thoughts as well. 

On the face of it, it does feel like breaking the API into their own endpoints is a good thing. Specially keeping in mind that agent might continue to grow in its feature set. Managing this single API to handle all the things in future might be problematic.

calvn replied:
The endpoint is a public API exposed by the agent, albeit a small one, so I think having it versioned to `/v1` is reasonable.

calvn replied:
The single endpoint supports clearing different parts the cache. You'd provide "request_path",  "token", "lease_id", or "all" as the type along with the value in the request body. We had it as a single endpoint in the RFC, but I'm working on getting it updated to be separate endpoints. I think that makes sense.

jefferai replied:
The problem with /v1/ is that if the agent ever uses a path that we want to use in Vault, now there is a clash. If you *must* version the api, then I'd suggest a pattern like `agent/v1` instead of `v1/agent`. Then there can be no confusion or overlap between a Vault path and an agent path.

Having seen the code for the cleaning endpoint I think having the single endpoint makes sense.

calvn replied:
Addressed in https://github.com/hashicorp/vault/pull/6257

jefferai commented on command/agent/cache/api_proxy.go:
```diff
@@ -0,0 +1,61 @@
+}
+
+func (ap *APIProxy) Send(ctx context.Context, req *SendRequest) (*SendResponse, error) {
+	client, err := api.NewClient(api.DefaultConfig())
```
I don't think you should create a new API client for every single call. This then requires creating a new HTTP transport/pool and so on. I think it should be sufficient to use Clone().

vishalnayak replied:
This was discussed between Chris and I and we had concluded on just using a new client. That thought was primarily based on the assumption that creating a new client is not going to be heavy. But, I'll work on Clone and see how that pans out.

jefferai replied:
If clients are using pooled connections you could pretty easily exhaust open file sockets if you are creating new connection pools for every connection. I think. So I do think this could be heavier than you think.

jefferai replied:
Lots of thinking in that comment. Thoughts.

vishalnayak replied:
Agreed. I hadn't thought of exhausting file sockets. This needs to be changed. I'm working on this.

vishalnayak replied:
Being tracked in https://github.com/hashicorp/vault/pull/6257/files

jefferai commented on command/agent/cache/api_proxy.go:
```diff
@@ -0,0 +1,61 @@
+		ap.logger.Error("failed to read request body", "error", err)
+		return nil, err
+	}
+	if resp.Body != nil {
```
ioutil.ReadAll is actually pretty expensive even if the body is nil. You're likely better off putting this check higher above so that you don't even call ioutil.ReadAll if the body is closed, and then just defer the close call.

jefferai commented on command/agent/cache/api_proxy.go:
```diff
@@ -0,0 +1,61 @@
+	if resp.Body != nil {
+		resp.Body.Close()
+	}
+	resp.Body = ioutil.NopCloser(bytes.NewBuffer(respBody))
```
What's the reason for this line?

jefferai replied:
You should probably consider switching the bytes.NewBuffer used in a bunch of these to bytes.NewReader to ensure that it can't be modified accidentally.

calvn replied:
Adressed in https://github.com/hashicorp/vault/pull/6257

jefferai commented on command/agent/cache/handler.go:
```diff
@@ -0,0 +1,155 @@
+			respondError(w, http.StatusInternalServerError, errors.New("failed to read request body"))
+		}
+		if r.Body != nil {
+			r.Body.Close()
```
Same comments here -- move the check up, and why is the next line there?

calvn replied:
Yep, we can move the nil check before the read, but same idea applies from the other comment as to why we're re-setting r.Body. Each of the layers might need to read the request, so we do that here before we send it to the first layer on `proxier.Send` so that it can simply read the request body using `SendRequest.RequestBody`. We reset the buffer using ioutil.NopCloser because when we get to the proxy layer the client uses the actual `r.Body`. We _could_ re-set it right before proxying, but we might as well reset it here.

jefferai commented on command/agent/cache/handler.go:
```diff
@@ -0,0 +1,155 @@
+	// Strip-off namespace related information from the request and get the
+	// relative path of the request.
+	_, path := deriveNamespaceAndRevocationPath(req)
+	if path == vaultPathTokenLookupSelf {
```
auth/token/lookup will use the request token if no other token is supplied, so you need to protect against it here too.

vishalnayak replied:
You are right. Thanks for caching this. This is being tracked here: https://github.com/hashicorp/vault/pull/6258.

jefferai commented on command/agent/cache/lease_cache.go:
```diff
@@ -0,0 +1,813 @@
+// Case 2: ns1/ /v1/leases/revoke -> ns1/, /v1/leases/revoke
+// Case 3: /v1/ns1/foo/bar  -> root/, /v1/ns1/foo/bar
+// Case 4: ns1/ /v1/foo/bar -> ns1/, /v1/foo/bar
+func deriveNamespaceAndRevocationPath(req *SendRequest) (string, string) {
```
I'm guessing this function goes away once https://github.com/hashicorp/vault/pull/6236 gets merged?

calvn replied:
We will still need this function in order to know what type of revocation request is being proxied.

jefferai commented on command/agent/cache/lease_cache.go:
```diff
@@ -0,0 +1,813 @@
+}
+
+// computeIndexID results in a value that uniquely identifies a request
+// received by the agent. It does so by SHA256 hashing the serialized request
```
I recommend using Blake2 instead of SHA256. You don't need the security and Blake is way faster.

vishalnayak replied:
Agreed. Fixing in https://github.com/hashicorp/vault/pull/6258/commits/f31ebb37c5d6aea26d0d9ad563d98d64aa5c65bb.

jefferai commented on command/agent/cache/lease_cache.go:
```diff
@@ -0,0 +1,813 @@
+		return "", fmt.Errorf("failed to serialize request: %v", err)
+	}
+
+	// Reset the request body after it has been closed by Write
```
This comment doesn't really make much sense, write isn't closing anything, the line below is.

calvn replied:
Write does actually close the body: https://golang.org/pkg/net/http/#Request.Write

> If Body is present, Content-Length is <= 0 and TransferEncoding hasn't been set to "identity", Write adds "Transfer-Encoding: chunked" to the header. Body is closed after it is sent.

I don't think we need to call req.Request.Body.Close() below.

jefferai replied:
Sure so if Write is closing it, the close below is spurious.

calvn replied:
Should be fixed in https://github.com/hashicorp/vault/pull/6237

jefferai commented on command/agent/cache/lease_cache.go:
```diff
@@ -0,0 +1,813 @@
+		req.Request.Body.Close()
+	}
+	req.Request.Body = ioutil.NopCloser(bytes.NewBuffer(req.RequestBody))
+
```
Won't this function not compute an equivalent index if the parameters are given in a different order?

calvn replied:
Yes, we settled on being acceptable for this first iteration (I forget if it was over Slack or one of the sync-ups). The expectation is that clients won't be switching back and forth on the ordering of the params that they request through agent since these are mainly going to be scripts or application services making the request.

jefferai replied:
The problem with that argument is that you assume JSON marshaling behavior is consistent from language to language or library to library. It may be, but it may not. Go lexically orders JSON keys, but there's a performance impact to doing so, so speed demon libs or implementations may simply randomly order the parameters.

As an example, even in Go such applications may use jsoniter instead of stdlib json, and in jsoniter sorting the keys is an option that is not turned on by default.

vishalnayak commented:
@jefferai While you continue to give comments on this PR, I'll be addressing the review comments in here https://github.com/hashicorp/vault/pull/6237/files.

jefferai commented on command/agent/cache/lease_cache.go:
```diff
@@ -0,0 +1,813 @@
+		c.logger.Error("failed to parse renewable param", "error", err)
+		return nil, err
+	}
+	if !secret.Renewable && !tokenRenewable {
```
You should still cache this for the lifetime of the secret/token, just not start a renewer process for it.

jefferai replied:
Maybe to put it another way, you may want to just remove this check, then in the cases below, add to index but don't start the renewer if it's not renewable.

vishalnayak replied:
Currently only the renewer goroutine when returning evicts entries from cache. The reason for designing this way is that we don't need to worry about invalidating the cache. There are no timers anywhere. But if we start caching non-renewable secrets, that won't happen. Although this seems like a very valid scenario to cache, we are simply not.

In fact the kind of items that are being cached (with v1 of this feature) are pretty scoped. Only the tokens that are created through the agent will be cached. This means that all the older tokens that transits through agent will be pass-through requests. Only the leases created by tokens that are already managed by the agent are cached. This gives more control on the items that are cached. However, this will most certainly change going forward and we'll add more and more cacheable scenarios will be added to the cache. Another goal with this approach was to ensure that the new feature will be a low touch thing, as in it won't disrupt the already existing tokens and lease management and is applicable only to new ones.

jefferai replied:
I would assume that revocation also evicts from the cache and stops any renewer activity?

But yeah I think a future direction here should be that anything with a lease gets cached regardless of renewability.

vishalnayak replied:
Yes, token or lease revocation results in the renewer getting killed, in-turn resulting in the eviction of respective entries from cache.

jefferai commented on command/agent/cache/lease_cache.go:
```diff
@@ -0,0 +1,813 @@
+
+	case secret.Auth != nil:
+		c.logger.Debug("processing auth response", "path", req.Request.URL.Path, "method", req.Request.Method)
+		isNonOrphanNewToken := strings.HasPrefix(req.Request.URL.Path, vaultPathTokenCreate) && resp.Response.StatusCode == http.StatusOK && !secret.Auth.Orphan
```
isNonOrphan will also be true when creating a token via the create-orphan endpoint or any role.

vishalnayak replied:
`isNonOrphanNewToken` will not be true when a token is created via create-orphan. Note that we are looking for non-orphan case here, not orphan cases. Orphan tokens will get cached. For non-orphan tokens, we want the renewer context to be derived from its parent token. The check seems fine unless I am missing something.

jefferai replied:
Ah, no. I'm missing something -- scrolling to the right. When I initially looked at this it cut off at just the right spot for me to not actually see the last part of the boolean expression.

That said, you should make that the first part -- both because it's a much faster check and it's the key part of the check.

Also -- why are the other parts of the expression necessary? Why not just gate purely on the Orphan check?

vishalnayak replied:
You are right. Other parts of the expression is unnecessary. Calvin will be removing it in https://github.com/hashicorp/vault/pull/6322.

jefferai commented on command/agent/cache/lease_cache.go:
```diff
@@ -0,0 +1,813 @@
+		}
+	}()
+
+	client, err := api.NewClient(api.DefaultConfig())
```
Clone a client...

vishalnayak replied:
https://github.com/hashicorp/vault/pull/6257

jefferai commented on command/agent/cache/lease_cache.go:
```diff
@@ -0,0 +1,813 @@
+
+		// Reset the base context
+		baseCtx, baseCancel := context.WithCancel(ctx)
+		c.baseCtxInfo = &ContextInfo{
```
This isn't thread-safe; you need to protect c.baseCtxInfo and all of its accesses with a mutex.

vishalnayak replied:
Mutexing baseCtxInfo may not be required. We set the `baseCtxInfo` while creation. We usually only read it. The only place it gets modified is when we cancel it and replace it with a new one, which happens over the API call. So, I am thinking that it should be fine. LMK if you think otherwise.

jefferai replied:
You are replacing it with a new one -- into a member variable. Which means you're modifying the member variable, and if that can happen in a request concurrent to a different request that reads it, you have a race condition. My guess is that concurrent requests are allowed in which case you definitely have a race here.

vishalnayak replied:
Fixed in https://github.com/hashicorp/vault/pull/6321

jefferai commented on command/agent/config/config.go:
```diff
@@ -22,6 +22,17 @@ type Config struct {
+	Listeners        []*Listener `hcl:"listeners"`
+}
+
+type Listener struct {
```
Can you instead just use the same Listener that we use in Vault? I think the parsing code below is basically a straight copy so it seems like a win to factor it out somewhere (e.g. helper/listener or so).

jefferai commented on command/server/listener.go:
```diff
@@ -72,7 +72,7 @@ func listenerWrapProxy(ln net.Listener, config map[string]interface{}) (net.List
 }
 
-func listenerWrapTLS(
+func ListenerWrapTLS(
```
Same comment basically -- I think we should move the listener stuff out somewhere common.
